File: src/main/java/com/anselumjuju/services/Workspaces.java
package com.anselumjuju.services;

import com.anselumjuju.lib.EnvConfig;
import com.anselumjuju.lib.AccessToken;
import com.anselumjuju.lib.Utils;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public class Workspaces {
    static final String ORG_ID = EnvConfig.ZOHO_ANALYTICS_ORG_ID;
    static final String ANALYTICS_URL = EnvConfig.ZOHO_AUTH_ANALYTICS_URL;
    static final String accessCode = AccessToken.getAccessToken();

    public static String createWorkspace() {
        String uniqueId = UUID.randomUUID().toString().replaceAll("-", "").substring(0, 10);
        String workspaceName = "workspace_" + uniqueId;

        JsonObject params = new JsonObject();
        params.addProperty("workspaceName", workspaceName);

        String url = ANALYTICS_URL + "/restapi/v2/workspaces?CONFIG=" + Utils.encode(params.toString());

        try (HttpClient client = HttpClient.newHttpClient()) {
            HttpRequest req = HttpRequest.newBuilder()
                    .uri(new URI(url))
                    .header("Authorization", "Zoho-oauthtoken " + accessCode)
                    .header("ZANALYTICS-ORGID", ORG_ID)
                    .POST(HttpRequest.BodyPublishers.noBody())
                    .build();

            HttpResponse<String> res = client.send(req, HttpResponse.BodyHandlers.ofString());

            Gson gson = new Gson();
            Map<String, Object> body = gson.fromJson(res.body(), new TypeToken<Map<String, Object>>() {
            }.getType());

            Map<String, Object> data = (Map<String, Object>) body.get("data");

            if (!body.get("status").equals("success")) {
                System.out.println("Error creating workspace: " + data.get("errorMessage"));
                return null;
            }

            return (String) data.get("workspaceId");
        } catch (Exception e) {
            System.out.println("Error creating workspace: " + e.getMessage());
            return null;
        }
    }

    public static List<String> getAllOwnedWorkspaces() {
        String url = ANALYTICS_URL + "/restapi/v2/workspaces";

        try (HttpClient client = HttpClient.newHttpClient()) {
            String accessCode = AccessToken.getAccessToken();
            HttpRequest req = HttpRequest.newBuilder()
                    .uri(new URI(url))
                    .header("Authorization", "Zoho-oauthtoken " + accessCode)
                    .GET()
                    .build();

            HttpResponse<String> response = client.send(req, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                System.out.println("Failed to fetch workspaces");
                return null;
            }

            Gson gson = new Gson();
            Map<String, Object> body = gson.fromJson(response.body(), new TypeToken<Map<String, Object>>() {
                    }.getType()
            );

            Map<String, Object> data = (Map<String, Object>) body.get("data");

            if (!body.get("status").equals("success")) {
                System.out.println("Error fetching workspaces: " + data.get("errorMessage"));
                return null;
            }

            List<Map<String, Object>> ownedWorkspaces = (List<Map<String, Object>>) data.get("ownedWorkspaces");

            List<String> workspaceIds = new ArrayList<>();
            for (Map<String, Object> workspace : ownedWorkspaces) {
                String name = (String) workspace.get("workspaceName");
                String id = (String) workspace.get("workspaceId");
                if (name != null && id != null && name.startsWith("workspace_"))
                    workspaceIds.add(id);
            }

            return workspaceIds;
        } catch (Exception e) {
            System.out.println("Error fetching workspaces: " + e.getMessage());
            return null;
        }
    }

    public static int deleteWorkspaces() {
        List<String> workspaces = getAllOwnedWorkspaces();
        if (workspaces == null || workspaces.isEmpty()) return 0;

        try (HttpClient client = HttpClient.newHttpClient()) {
            String accessCode = AccessToken.getAccessToken();
            List<CompletableFuture<Boolean>> futures = new ArrayList<>();

            for (String workspaceId : workspaces)
                futures.add(deleteWorkspaceById(client, accessCode, workspaceId));

            CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();

            int deletedCount = 0;
            for (CompletableFuture<Boolean> future : futures)
                if (future.join()) deletedCount++;

            return deletedCount;
        } catch (Exception e) {
            System.out.println("Error deleting workspaces: " + e.getMessage());
            return 0;
        }
    }

    public static CompletableFuture<Boolean> deleteWorkspaceById(HttpClient client, String accessCode, String workspaceId) throws Exception {
        String url = ANALYTICS_URL + "/restapi/v2/workspaces/" + workspaceId;

        HttpRequest req = HttpRequest.newBuilder()
                .uri(new URI(url))
                .header("Authorization", "Zoho-oauthtoken " + accessCode)
                .header("ZANALYTICS-ORGID", ORG_ID)
                .DELETE()
                .build();

        return client.sendAsync(req, HttpResponse.BodyHandlers.ofString())
                .thenApply(res -> res.statusCode() == 200)
                .exceptionally(ex -> {
                    System.err.println("Failed to delete workspace " + workspaceId + ": " + ex.getMessage());
                    return false;
                });
    }
}

File: backend/src/services/workspaces.js
import {env} from '../lib/env.js';
import {encode} from '../lib/utils.js';
import {getValidAccessToken} from './accessToken.js';

export async function createWorkspace() {
  const uniqueId = crypto.randomUUID().replaceAll('-', '').slice(0, 5);
  const uniqueTimestamp = String(Date.now());
  const workspaceName = `workspace_${uniqueId}_${uniqueTimestamp}`;
  const params = encode(JSON.stringify({workspaceName}));
  const url = `${env.ZOHO_AUTH_ANALYTICS_URL}/restapi/v2/workspaces?CONFIG=${params}`;

  try {
    const accessToken = await getValidAccessToken();
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `Zoho-oauthtoken ${accessToken}`,
        'ZANALYTICS-ORGID': env.ZOHO_ANALYTICS_ORG_ID,
      },
    });

    const body = await response.json();
    if (body.status !== 'success') {
      console.log(`Error creating workspace: ${body?.data?.errorMessage || 'Unknown error'}`);
      return null;
    }

    return body?.data?.workspaceId || null;
  } catch (error) {
    console.log(`Error creating workspace: ${error.message}`);
    return null;
  }
}

export async function getAllOwnedWorkspaces() {
  const url = `${env.ZOHO_AUTH_ANALYTICS_URL}/restapi/v2/workspaces`;

  try {
    const accessToken = await getValidAccessToken();
    const response = await fetch(url, {headers: {Authorization: `Zoho-oauthtoken ${accessToken}`}});
    if (response.status !== 200) return null;

    const body = await response.json();
    if (body.status !== 'success') return null;

    const ownedWorkspaces = body?.data?.ownedWorkspaces || [];
    return ownedWorkspaces
      .filter((workspace) => typeof workspace.workspaceName === 'string' && workspace.workspaceName.startsWith('workspace_'))
      .map((workspace) => workspace.workspaceId)
      .filter(Boolean);
  } catch (error) {
    console.log(`Error fetching workspaces: ${error.message}`);
    return null;
  }
}

export async function deleteWorkspaces() {
  const workspaces = await getAllOwnedWorkspaces();
  if (!workspaces || workspaces.length === 0) return 0;

  try {
    const accessToken = await getValidAccessToken();
    const results = await Promise.all(workspaces.map(async (workspaceId) => deleteWorkspaceById(accessToken, workspaceId)));
    return results.filter(Boolean).length;
  } catch (error) {
    console.log(`Error deleting workspaces: ${error.message}`);
    return 0;
  }
}

async function deleteWorkspaceById(accessToken, workspaceId) {
  const url = `${env.ZOHO_AUTH_ANALYTICS_URL}/restapi/v2/workspaces/${workspaceId}`;

  try {
    const response = await fetch(url, {
      method: 'DELETE',
      headers: {
        Authorization: `Zoho-oauthtoken ${accessToken}`,
        'ZANALYTICS-ORGID': env.ZOHO_ANALYTICS_ORG_ID,
      },
    });

    return response.status === 204;
  } catch (error) {
    console.log(`Failed to delete workspace ${workspaceId}: ${error.message}`);
    return false;
  }
}
